exercises/bank-account/canonical-data.json

```diff
@@ -0,0 +1,303 @@
+  "comments": [
+    "For this exercise it is possible to implement concurrency tests",
+    "but for language tracks that do not support concurrency, the tests",
+    "are written to be run in a single thread.",
+    "But for language tracks that do support concurrency, it is possible",
+    "to write tests that run concurrently."
+  ],
+  "cases": [
+    {
+      "uuid": "983a1528-4ceb-45e5-8257-8ce01aceb5ed",
+      "description": "Newly opened account has zero balance",
+      "property": "bankAccount",
+      "input": {
+        "operations": [
+          {
+            "operation": "open"
+          },
+          {
+            "operation": "amount",
+            "expected": 0
+          }
+        ]
+      },
+      "expected": {}
+    },
+    {
+      "uuid": "e88d4ec3-c6bf-4752-8e59-5046c44e3ba7",
+      "description": "Can deposit money",
+      "property": "bankAccount",
+      "input": {
+        "operations": [
+          {
+            "operation": "open"
+          },
+          {
+            "operation": "deposit",
+            "value": 100
+          },
+          {
+            "operation": "amount",
+            "expected": 100
+          }
+        ]
+      },
+      "expected": {}
+    },
+    {
+      "uuid": "3d9147d4-63f4-4844-8d2b-1fee2e9a2a0d",
+      "description": "Can deposit money sequentially",
+      "property": "bankAccount",
+      "input": {
+        "operations": [
+          {
+            "operation": "open"
+          },
+          {
+            "operation": "deposit",
+            "value": 100
+          },
+          {
+            "operation": "deposit",
+            "value": 50
+          },
+          {
+            "operation": "amount",
+            "expected": 150
+          }
+        ]
+      },
+      "expected": {}
+    },
+    {
+      "uuid": "08f1af07-27ae-4b38-aa19-770bde558064",
+      "description": "Can withdraw money",
+      "property": "bankAccount",
+      "input": {
+        "operations": [
+          {
+            "operation": "open"
+          },
+          {
+            "operation": "deposit",
+            "value": 100
+          },
+          {
+            "operation": "withdraw",
+            "value": 50
+          },
+          {
+            "operation": "amount",
+            "expected": 50
+          }
+        ]
+      },
+      "expected": {}
+    },
+    {
+      "uuid": "6f6d242f-8c31-4ac6-8995-a90d42cad59f",
+      "description": "Can withdraw money sequentially",
+      "property": "bankAccount",
+      "input": {
+        "operations": [
+          {
+            "operation": "open"
+          },
+          {
+            "operation": "deposit",
+            "value": 100
+          },
+          {
+            "operation": "withdraw",
+            "value": 80
+          },
+          {
+            "operation": "withdraw",
+            "value": 20
+          },
+          {
+            "operation": "amount",
+            "expected": 0
```

Comment:
If we're defining `operations` to have an operand / function call name, and optional argument of `value`, and an optional return expectation of `expected`, I think we should move `error` also into an operation expectation.

This also solves the potential ordering issue @petertseng mentions, and makes it easier to add interesting concurrency tests down the line.

Reply:
Thanks for your feedback, do you think writing it like this makes sense?

```
          {
            "operation": "deposit",
            "value": -50,
            "error": "amount must be greater than 0"
          }
```

Reply:
Perhaps still nest it in `expected` like before? @petertseng what do you think?

Reply:
Is there a reason to prefer one over the other? Since I'm not currently aware of one, it seems we are empowered to make either choice. If a reason presents itself, then we would go with that one. For example, someone might say "all other `error` are currently under `expected`, and it is important for this to remain true because (they give a reason here)". You will notice that since I didn't have something to insert in the blank that I didn't put forth that reason.

Reply:
Since no one has written any test generator for this format yet, I agree, there is no reason from that point of view right now to pick one or the other.

@meatball133 I'll leave it up to you unless someone else comes up with a good reason.

Reply:
I find the empty `expected` sibling to `input` quite confusing.

What if we have some sort of flag that determines which operation produces the expected result?

```json
    {
      "uuid": "983a1528-4ceb-45e5-8257-8ce01aceb5ed",
      "description": "Newly opened account has zero balance",
      "property": "bankAccount",
      "input": {
        "operations": [
          {
            "operation": "open"
          },
          {
            "operation": "amount",
            "result": true
          }
        ]
      },
      "expected": 0
    },
...
    {
      "uuid": "f9facfaa-d824-486e-8381-48832c4bbffd",
      "description": "Checking balance of closed account throws error",
      "property": "bankAccount",
      "input": {
        "operations": [
          {
            "operation": "open"
          },
          {
            "operation": "close"
          },
          {
            "operation": "amount",
            "result": true
          }
        ]
      },
      "expected": {
        "error": "account not open"
      }
    }
```

Reply:
Or, just remove the empty outer "expected"

Reply:
I think the outer one should go away indeed.

Reply:
I think the design you suggested Gleen could just make stuff very complicated for generators.
Say you would want to have 2 expect?, then it would be quite hard to implement that with such a system.

And removing expect seems to go against the design scheme.

Reply:
Ah, right. I'm now thinking that the input should be an array of operations without any "result" property, and the expected result is the result of the last operation.

Peter's previous comment "If there were ever a case in the future where an operation other than the last one should cause the error": if I were writing unit tests, I wouldn't be performing any more operations following an operation that produces an expected error (unless we're testing the behaviour of the system following an error, in which case we'd explicitly catch the error)

I propose we keep it simple: we'll just expect the results of the last operation.

Would this be obvious to people implementing the test suite for a track? Would we need a comment explicitly stating it?

Reply:
I would always add a comment for a future reader.

Reply:
> unless we're testing the behaviour of the system following an error

This type of test could be quite valuable if one of the desired teaching points of this exercise is to ensure that the system remains in a usable state after an erroneous operation. Ideally that decision wouldn't need to be made now, but I do regret to say that since it might have an effect on how some future tests might need to be written, it seems it's necessary to at least plan for it. If there is no intent at all to have this as a teaching point, then all is well and we can end the discussion here, and reading the rest of this comment is unnecessary.

But if error handling is one of the desired teaching points of this exercise, then just as an example, consider a test that opens an account, deposits a negative amount (which is not allowed), and then checks that the balance is still zero following that erroneous operation:

1. Operations other than the last one could have an `expected`, with the outer `expected` still referring to the very last operation:

```json
    {
      "uuid": "785f57e4-7c32-402e-a118-e3bfd4facf75",
      "description": "Bad deposit",
      "property": "bankAccount",
      "input": {
        "operations": [
          {
            "operation": "open"
          },
          {
            "operation": "deposit",
            "value": -50,
            "expected": {"error": "amount must be greater than 0"}
          },
          {
            "operation": "amount"
          }
        ]
      },
      "expected": 0
    }
```

2. Or, each operation could have its own `expected` and the outer one remains empty.


```json
    {
      "uuid": "785f57e4-7c32-402e-a118-e3bfd4facf75",
      "description": "Bad deposit",
      "property": "bankAccount",
      "input": {
        "operations": [
          {
            "operation": "open"
          },
          {
            "operation": "deposit",
            "value": -50,
            "expected": {"error": "amount must be greater than 0"}
          },
          {
            "operation": "amount"
            "expected": 0
          }
        ]
      },
      "expected": {}
    }
```

I have no reason to prefer one or the other, but if someone does, it would be good to mention the reason for that preference right now.

I also wouldn't consider having such tests a necessary condition to have this pull request be Approved (in the GitHub sense of the word), but I'll phrase it as: It wouldn't be surprising to see an additional pull request from me some time after this pull request is merged if this is indeed a desired teaching point.

Reply:
I hear what you're saying about the teaching point.

Of these options, I prefer #1. That indicates that we want to produce *that result* from the *entire sequence* of operations. Contrast to #2 where the actual expected result of the test case is ambiguous.

I think we'd want to give some hint to track implementors that in this situation, the erroring operation has in a non-fatal error or that the error is to be trapped. Perhaps adding a (new) scenario to the case or adding a `"fatal": false` property to the `expected` object for that operation.

Reply:
Sorry for taking a while, I will replay later today. Promise, aslong as I dont get a power outage but that usually doesnt happen.

Reply:
Okay I have read through the feedback. I think we can go with method one. But there are no current test cases which has any other operations in the middle so the fatal flag doesn't need to be applied anywhere. Unless you guys come up with some new tests. But I can convert the tests to suggestion 1. I can also write a comment about how we think about this design.